core/src/main/scala/kafka/server/AbstractRequestHandler.scala

```diff
@@ -0,0 +1,153 @@
+package kafka.server
+
+import java.lang.{Byte => JByte}
+import java.util.Collections
+
+import kafka.network.RequestChannel
+import kafka.security.authorizer.AclEntry
+import kafka.server.QuotaFactory.QuotaManagers
+import kafka.utils.Logging
+import org.apache.kafka.common.acl.AclOperation
+import org.apache.kafka.common.errors.ClusterAuthorizationException
+import org.apache.kafka.common.network.Send
+import org.apache.kafka.common.requests.{AbstractRequest, AbstractResponse, RequestContext}
+import org.apache.kafka.common.resource.Resource.CLUSTER_NAME
+import org.apache.kafka.common.resource.ResourceType.CLUSTER
+import org.apache.kafka.common.resource.{PatternType, Resource, ResourcePattern, ResourceType}
+import org.apache.kafka.common.utils.{Time, Utils}
+import org.apache.kafka.server.authorizer.{Action, AuthorizationResult, Authorizer}
+
+import scala.jdk.CollectionConverters._
+
+/**
+ * Base class for request handlers. Provides common functionality around throttling and error handling
+ */
+abstract class AbstractRequestHandler(val requestChannel: RequestChannel,
```

Comment:
It seems to me that having this as a class instead of a abstract class would be cleaner. Any reason not to do that?

Reply:
Yea, I was on the fence with an abstract class as well. Doing this as a helper class is nice and clean for ControllerApis, but adds another dependency on KafkaApis (since KafkaApis has other usages of Authorizer and QuotaManager). That said, I went ahead and pull this into a separate class. Let me know what you think

Reply:
Thanks, @mumrah !  I like the idea of having a separate class for these utility functions.  It's cleaner and simpler to understand than using inheritance, I think.

Is there a reason why `RequestChannelHelper` needs to be in the constructor of `KafkaApis` and `ControllerApis`?  As far as I can tell, all the things `RequestChannelHelper` needs are already being passed in to those classes, so can't it just create an instance of this internally?

Even if we take it out of the constructor, though, I see two problems with the current code layout.  One is that some of the `RequestChannelHelper` functions don't really have anything to do with request channels.  So `RequestChannelHelper#authorize` is doing something that has nothing to do with a request channel... confusing.  The other is that usually when we have a class with some data members, it "owns" those members in some sense and is responsible for managing them.  That's clearly not the case for `RequestChannelHelper`, and I think that's also confusing.

Maybe this would be simpler as a set of static utility methods (or in Scala-land, a "companion object without state")?  We could move `authorize`, `authorizerClusterOperations`, and `authorizedOperations` into `AuthorizerUtils`.  After all, they're clearly utility functions that have to do with the authorizer.  And then maybe have a `RequestChannelUtils` for the other functions?

Reply:
I agree that `RequestChannelHelper` doesn't represent well what this class is doing. I don't think the class has to own the members though given that the members are services. It can help avoid a situation where you have a really large number of parameters for every method. I think there are two possible paths:
1. Go with the approach Colin suggested. I think this makes sense if these methods make sense in isolation.
2. Have a ApisUtils class that is focused on providing utility methods to bridge what the *Apis class need and what `RequestChannel` and `Authorizer` provide.

Reply:
A utility/helper class is more like what I was going for. I think keeping them in a utility class has the least impact on KafkaApis while still having some benefit.

If a class creates an object, I think it's fair to assume it should assume responsibility for it. But if it's passed in as a dependency/service it seems fine for the class to just make use of it.

Also agree that `RequestChannelHelper` is a bad name. `ApiRequestHandlerHelper` would have been better, but i like `ApisUtils` even more.

Reply:
I still feel like this code could be structured better.  But let's merge this PR for now so we can unblock the other parts of the project.  LGTM.